contracts/Joint.sol

```diff
@@ -292,369 +358,418 @@ abstract contract Joint is ySwapper {
+     * @param _token_in, address of the token to swap from
+     * @param _token_to, address of the token to swap to
+     * @return address array of the swap path to follow
+     */
     function getTokenOutPath(address _token_in, address _token_out)
         internal
         view
         returns (address[] memory _path)
     {
-        bool is_weth =
-            _token_in == address(WETH) || _token_out == address(WETH);
-        bool is_internal =
-            (_token_in == tokenA && _token_out == tokenB) ||
-                (_token_in == tokenB && _token_out == tokenA);
-        _path = new address[](is_weth || is_internal ? 2 : 3);
+        bool isReferenceToken = _token_in == address(referenceToken) ||
+            _token_out == address(referenceToken);
+        bool is_internal = (_token_in == tokenA && _token_out == tokenB) ||
+            (_token_in == tokenB && _token_out == tokenA);
+        _path = new address[](isReferenceToken || is_internal ? 2 : 3);
         _path[0] = _token_in;
-        if (is_weth || is_internal) {
+        if (isReferenceToken || is_internal) {
             _path[1] = _token_out;
         } else {
-            _path[1] = address(WETH);
+            _path[1] = address(referenceToken);
             _path[2] = _token_out;
         }
     }
 
     function getReward() internal virtual;
 
-    function depositLP() internal virtual;
+    function depositLP() internal virtual {}
+
+    function withdrawLP() internal virtual {}
 
-    function withdrawLP() internal virtual;
+    struct tokenAmount {
+        address token;
+        uint256 amount;
+    }
 
-    function swapReward(uint256 _rewardBal)
+    /*
+     * @notice
+     *  Function available internally swapping amounts necessary to swap rewards
+     * @return tokenAmount array of the swap path followed
+     */
+    function swapRewardTokens()
         internal
         virtual
-        returns (address, uint256)
+        returns (tokenAmount[] memory)
     {
-        if (reward == tokenA || reward == tokenB || _rewardBal == 0) {
-            return (reward, 0);
-        }
-
-        if (tokenA == WETH || tokenB == WETH) {
-            return (WETH, sellCapital(reward, WETH, _rewardBal));
-        }
-
-        // Assume that position has already been liquidated
-        (uint256 ratioA, uint256 ratioB) =
-            getRatios(balanceOfA(), balanceOfB(), investedA, investedB);
-        if (ratioA >= ratioB) {
-            return (tokenB, sellCapital(reward, tokenB, _rewardBal));
+        tokenAmount[] memory _swapToAmounts = new tokenAmount[](
+            rewardTokens.length
+        );
+        for (uint256 i = 0; i < rewardTokens.length; i++) {
+            address reward = rewardTokens[i];
+            uint256 _rewardBal = IERC20(reward).balanceOf(address(this));
+            // If the reward token is either A or B, don't swap
+            if (reward == tokenA || reward == tokenB || _rewardBal == 0) {
```

Comment:
as tokenB (and tokenA) is used several times in each loop and it is a SLOAD, I would cache both _tokenA and _tokenB as a local variable and read from it (for gas savings)

Reply:
True, implementing it